wgsl/index.bs

```diff
@@ -632,16 +633,71 @@ module header and each usage of a GLSL method will add the appropriate
   </xmp>
 </div>
 
+## Module Constants ## {#module-constants}
+
+A *module constant* declares a name for a value, outside of all function declarations.
+The name is available for use after the end of the declaration,
+until the end of the WGSL program.
+
+When the declaration has no attributes, the value is the expression in the declaration.
+
+<div class='example' heading='Module constants'>
+  <xmp>
+    const golden : f32 = 1.61803398875;       # The golden ratio
+    const e2 : vec3<i32> = vec3<i32>(0,1,0);  # The second unit vector for three dimensions.
+  </xmp>
+</div>
+
+When the declaration uses the `constant_id` attribute,
+the constant is *pipeline-overridable*. In this case:
+
+  * The type must one of the [[#scalar-types]].
+  * The attribute's literal operand is known as the *pipeline constant ID*,
+    and must be a non-negative integer value representable in 32 bits.
+  * The application can specify its own value for the name at pipeline-creation time.
+    The pipeline creation API accepts a mapping from the pipeline constant ID
+    to a value of the constant's type.
+    If the mapping does not have an entry for the pipeline constant ID, then the expression
+    in the declaration is used as the value.
+  * Pipeline constant IDs must be unique within the WGSL program: Two module constants
+    must not use the same pipeline constant ID.
+
+<div class='example' heading='Module constants, pipeline-overrideable'>
+  <xmp>
+    [[constant_id 0]]    const has_point_light : bool = true;      # Algorithmic control
+    [[constant_id 1200]] const specular_param : f32 = 2.3;         # Numeric control
+  </xmp>
+</div>
+
+When a variable or feature is used within control flow that depends on the
+value of a constant, then that variable or feature is considered to be used by the
+program.
+This is true regardless of the value of the constant, whether that value
+is the one from the constant's declaration or from a pipeline override.
+
+<pre class='def'>
+global_constant_decl
+  : global_const_decoration_list? CONST variable_ident_decl EQUAL const_expr
+
+global_const_decoration_list
+  : ATTR_LEFT global_const_decoration ATTR_RIGHT
+
+global_const_decoration
+  : CONSTANT_ID INT_LITERAL
+</pre>
+
+Issue(dneto): The WebGPU pipeline creation API must specify how API-supplied values are mapped to
```

Comment:
We have the concept of host-accessible types in the spec, right? Can we just say the type of a constant has to be a host-accessible scalar?

Reply:
That works

Reply:
Hmm. That removes the ability to set booleans.   It's just as easy to say that when setting a bool constant, supply a 32-bit integer value, with 0 mapping to false, and true otherwise.  It depends on the shape of the API whether we even need that detail or not.